### ISSUE TITLE
singularity uses a large amount of memory while unpacking docker image

### ISSUE THREAD

**Issue by soichih** [apptainer/singularity#5295]
I am running singularity 3.5.2 on IU BigRed3 cluster (SLES 12 SP3). 

When I run a large docker container (~5G) with singularity using  `docker://poldracklab/fmriprep:1.5.10` format, it often gets killed by the cluster due to OOM while unpacking layer.

```
+ singularity exec -e docker://poldracklab/fmriprep:1.5.10 /usr/local/miniconda/bin/fmriprep --notrack --resource-monitor --skip-bids-validation --md-only-boilerplate --stop-on-first-crash --use-plugin=multi_proc.yml --output-spaces T1w MNI152NLin2009cAsym --force-bbr --use-syn-sdc --fs-license-file=license.txt --skull-strip-template=NKI --work-dir=fmripworkdir --participant_label=sub-TTTEMPSUB fmripinput fmripout participant
[34mINFO:   [0m Converting OCI blobs to SIF format
[34mINFO:   [0m Starting build...
Getting image source signatures
Copying blob sha256:c832082614738021cf51d68e84457a45a3f85dbe0ddbb93000c5dc5aae77fe93
Copying blob sha256:6e1a85c1d66a18674bdb5cee0a7d6b294aa4f192
...
...

2020/05/15 16:53:05  info unpack layer: sha256:6ad2e8634be5c7463703047dd6cea868ea61f12ec61164ae591147e04fece83e
2020/05/15 16:53:05  info unpack layer: sha256:160eb4504b1bea3d13f279c215fc998c2f241cd3c760b967effb35b05d63cf5b
2020/05/15 16:53:05  warn xattr{etc/gshadow} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/05/15 16:53:05  warn xattr{/tmp/rootfs-d1f18631-96ed-11ea-877e-000101000058/etc/gshadow} destination filesystem does not support xattrs, further warnings will be suppressed
2020/05/15 16:53:10  info unpack layer: sha256:3c248cebabad93e4291aba65eeaa46fb9341a1ca3cc1c230e0ead5f4958814d5
slurmstepd: error: Detected 1 oom-kill event(s) in step 417163.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
slurmstepd: error: get_exit_code task 0 died by signal: 9

```

I am allocating 14gb of memory for this job.

```
#!/bin/bash
#PBS -l nodes=1:ppn=8,walltime=9:00:00,vmem=14gb
```

The container itself uses much less than 14gb, so I'd rather not increase the memory requirement, but according to top command, `mksquashfs` uses up to 20gb of memory while converting the image to SIF.

![image](https://user-images.githubusercontent.com/923896/82097260-8cee1a80-96d0-11ea-81fe-51771428d8e2.png)

Is this expected that singularity uses a lot of memory while creating the SIF image? Is there a workaround to use less memory?

**Comment by dtrudg** [comment]
This is expected and is related to the SIF creation. It is not specific to docker sources, and it is the `mksquashfs` tool provided by the distro rather than Singularity itself which is using the memory. `mksquashfs` which is called by Singularity will use all cores by default, and up to 25% of system memory. From the man page:

```
       -processors NUMBER
           Use NUMBER processors. By default will use number of processors available.

       -mem SIZE
           Use SIZE physical memory. Optionally K or M can be used as a suffix for kilobytes or megabytes, respectively. Default 25% of memory.
```

In 3.6.0 there will be a way to limit the processors and memory usage through options in `singularity.conf` that pass the above flags to `mksquashfs` when it is invoked.

See this PR: https://github.com/hpcng/singularity/pull/5003

**Comment by dtrudg** [comment]
Sorry - I should have mentioned that the `-mem` limit is only available on newer squashfs versions. On older versions you will only be able to limit `-processors`, so you can limit memory usage by proxy by reducing the processors used - or allocate more memory for the job when creating very large containers.

**Comment by soichih** [comment]
@dctrud Thank you for your reply. Am I correct to understand that we will not be able to limit the mksquashfs memory usage for singularity 3.5.2? I will ask our sysadmin to upgrade to 3.6. 

Also, is there a way to limit the mksquashfs memory usage via singularity command line option, or ENV?

**Comment by dtrudg** [comment]
You cannot limit the usage on Singularity 3.5.2 - correct. 3.6.0 is not released yet, but there is a release candidate for testing.

There is no way to limit via option or ENV. If you would like to request this please open a specific issue asking for this. PRs / code contributions are always welcome too.

Cheers!

**Comment by dtrudg** [comment]
This is now available to set in `singularity.conf` and working on `3.6.1`

**Comment by soichih** [comment]
@dtrudg Is this option still available for singularity 3.7.2? Is there a way to set this via ENV?

I have 127 cores on our login node and I am running into the following error message.

```
INFO:    Converting OCI blobs to SIF format
FATAL:   Unable to handle docker://brainlife/cli uri: while building SIF from layers: unable to create new build: while ensuring correct compression algorithm: while creating squashfs: create command failed: exit status 1: FATAL ERROR:Failed to create thread
FATAL ERROR: Out of memory (frag_thrd)
```

**Comment by DrDaveD** [comment]
Hi Soichi.  dtrudg no longer follows this singularity, because sylabs made a fork they call singularity-ce.  If you need to reach him, make a new issue at https://github.com/sylabs/singularity.
